Node.js client (#233)
infinispan/infinispan.github.io: _posts/2021-06-29-js-client.adoc

---
layout: blog
title: Infinispan Node.js client supports authentication
permalink: /blog/:year/:month/:day/infinispan-js-client
date: '2021-06-29T12:00:00.000-00:00'
author: karesti
tags: [ "hotrod", "Node.js", "client", "javascript", "release" ]
---

== Node.js Client 0.9.0

> Infinispan Node.js client 0.9.0 has been released last week.
> Until now, this client supported connections with an unsecured Infinispan Server.
> Since version 0.9, the client supports different authentication mechanisms.

Reviewer: ```suggestion
Infinispan Node.js client version 0.9.0 was released last week with added support for different SASL authentication mechanisms.
Up to now, our Node.js client could connect to Infinispan Server security realms with disabled authentication.
```

=== DIGEST-MD5
Uses the MD5 hashing algorithm in addition to nonces to encrypt credentials.

=== SCRAM
Uses salt values in addition to hashing algorithms and nonce values to encrypt credentials. Hot Rod endpoints support SCRAM-SHA-1, SCRAM-SHA-256, SCRAM-SHA-384, SCRAM-SHA-512 hashing algorithms, in order of strength.

=== EXTERNAL
Uses client certificates to provide valid identities to Infinispan Server and enable encryption.

=== OAUTHBEARER
Uses tokens obtained via an OAuth 2.0 provider to securely connect to Infinispan Server.

=== PLAIN:
Sends credentials in plain text (unencrypted) over the wire in a way that is similar to HTTP BASIC authentication.

WARNING: To secure Infinispan credentials, you should use PLAIN authentication only in combination with TLS encryption.

== Run the Infinispan Server

.Run the Infinispan Server with Docker or Podman
[source,bash]
----
docker run -it -p 11222:11222 -e USER="admin" -e PASS="password" quay.io/infinispan/server:12.1

podman run -it -p 11222:11222 -e USER="admin" -e PASS="password" --net=host quay.io/infinispan/server:12.1
----

> IMPORTANT: If you are using Docker for Mac, there is known limitation. You will need to https://infinispan.org/download/[download

Reviewer: ```suggestion
IMPORTANT: If you are using Docker for Mac, there is a known limitation. You will need to https://infinispan.org/download/[download
```

the server distribution] and run the server manually.

> .Run the Infinispan Server manually from the downloader folder

Reviewer: ```suggestion
.Run the Infinispan Server from the file system
```

[source,bash]
----
./bin/cli.sh user create admin -p password
./bin/server.sh
----

> == Create a cache from the Infinispan Console

Reviewer: ```suggestion
== Create a cache from Infinispan Console
```

Access the Infinispan Console in http://localhost:11222/[http://localhost:11222/] and create a text based
cache, named it 'my-cache' with the provided configuration.

.Connect to Infinispan
[source,json]
----
{
  "distributed-cache": {
    "mode": "SYNC",
    "encoding": {
      "media-type": "text/plain"
    },
    "statistics": true
  }
}
----

== Use the Node.js client in your application

> Add the dependency in your project.

Reviewer: ```suggestion
Add the dependency to your project.
```

.package.json
[source,json]
----
"dependencies": {
"infinispan": "^0.9.0"
}
----

> Use the Infinispan Node.js Client authenticated, and check the created cache entry from the console.

Reviewer: ```suggestion
Configure the Infinispan Node.js client to connect with authentication and then check the created cache entry from the console.
```

.application.js
[source,javascript]
----
var connected = infinispan.client({port: 11222, host: '127.0.0.1'},
{
    cacheName: 'my-cache',
    authentication: {
        enabled: true,
        saslMechanism: 'DIGEST-MD5',
        userName: 'admin',
        password: 'password'
    }
});

connected.then(function (client) {
    return client.put('key', 'value')
    .finally(function() {
    return client.disconnect();
  });
});
----

== To go further

> The full client documentation is now available in the

Reviewer: ```suggestion
Full client documentation is now available in the
```

> https://infinispan.org/docs/hotrod-clients/js/latest/js_client.html[Infinispan documentation].

Reviewer: ```suggestion
https://infinispan.org/docs/hotrod-clients/js/latest/js_client.html[Hot Rod JS Client Guide].
```
